packages/devtools-map-bindings/src/parser.js

```diff
@@ -0,0 +1,297 @@
+        case "param":
+          _bindings[n] = nameRefs.refs.map(location => {
+            return fromBabelLocation(location, sourceId);
+          });
+          break;
+      }
+      return _bindings;
+    }, Object.create(null));
+    return {
+      start: fromBabelLocation(scope.loc.start, sourceId),
+      end: fromBabelLocation(scope.loc.end, sourceId),
+      type: scope.type.toLowerCase(),
+      bindings,
+      children: toParsedScopes(scope.children, sourceId)
+    };
+  });
+}
+
+/**
+ * Creates at visitor for babel-traverse that will parse/extract all bindings
+ * information from the source. See also findScopes to perform lookup of the
+ * scope information for specific location.
+ */
+function createParseJSScopeVisitor(sourceId: SourceId): ParseJSScopeVisitor {
+  let parent: TempScope = createTempScope("Global", null, null);
+  let savedParents: WeakMap<NodePath, TempScope> = new WeakMap();
+  const traverseVisitor = {
+    enter(path: NodePath) {
+      const tree = path.node;
+      const location = path.node.loc;
+      if (path.isProgram()) {
+        savedParents.set(path, parent);
+        parent = createTempScope("Script", parent, location);
+        return;
+      }
+      if (
+        path.isFunctionDeclaration() ||
+        path.isFunctionExpression() ||
+        path.isArrowFunctionExpression()
+      ) {
+        savedParents.set(path, parent);
+        const scope = createTempScope("Function", parent, location);
+        if (isNode(tree.id, "Identifier")) {
+          const functionName = { type: "fn", refs: [] };
+          getFunctionScope(parent).names[tree.id.name] = functionName;
+          scope.names[tree.id.name] = functionName;
+        }
+        tree.params.forEach(param => {
+          parseDeclarator(param, scope, "param");
+        });
+        parent = scope;
+        return;
+      }
+      if (path.isForInStatement() || path.isForStatement()) {
+        const init = tree.init || tree.left;
+        if (
+          isNode(init, "VariableDeclaration") &&
+          (init.kind === "let" || init.kind === "const")
+        ) {
+          savedParents.set(path, parent);
+          parent = createTempScope("For", parent, location);
+        }
+        return;
+      }
+      if (path.isCatchClause()) {
+        savedParents.set(path, parent);
+        const scope = createTempScope("Catch", parent, location);
+        parseDeclarator(tree.param, scope, "param");
+        parent = scope;
+        return;
+      }
+      if (path.isBlockStatement()) {
+        const hasLet = tree.body.some(node => {
+          if (!isNode(node, "VariableDeclaration")) {
+            return false;
+          }
+          return node.kind === "let" || node.kind === "const";
+        });
+        if (hasLet) {
+          savedParents.set(path, parent);
+          parent = createTempScope("Block", parent, location);
+        }
+        return;
+      }
+      if (path.isVariableDeclaration()) {
+        const hoistAt =
```

Comment:
neat. I imagine this might be one of the more interesting considerations because babel is free to add other hoisting rules in the future:

perhaps we want to try looking for the ref:

```js
path.findParent((path) => path.scope.bindings.include(?));
```

NOTE: just guessing based on reading the code, i could be mis-reading this